eth/protocols/snap/gentrie.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package snap
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// genTrie interface is used by the trie to generate merkle tree nodes based
+// on a received batch of states.
+type genTrie interface {
+	// update inserts the state into generator trie.
+	update(key, value []byte) error
+
+	// commit flushes the leftover nodes produced in the trie into database.
+	// The nodes on right boundary won't be committed unless this function
+	// is called. The flag complete should be set to true if there are more
+	// items on the right side.
+	//
+	// This function must be called before flushing database batch.
+	commit(complete bool) common.Hash
+}
+
+// pathTrie is a wrapper over the stackTrie, incorporating numerous additional
+// logics to handle the semi-completed trie and potential leftover dangling
+// nodes in the database. It is utilized for constructing the merkle tree nodes
+// in path mode during the snap sync process.
+type pathTrie struct {
+	owner common.Hash     // identifier of trie owner, empty for account trie
+	tr    *trie.StackTrie // underlying raw stack trie
+	first []byte          // the path of first written node
+	last  []byte          // the path of last written node
+
+	// Flag whether the nodes on the left boundary are skipped for committing.
+	// If it's set, then nodes on the left boundary are regarded as incomplete
+	// due to potentially missing left children.
+	noLeftBound bool
+	db          ethdb.KeyValueReader
+	batch       ethdb.Batch
+}
+
+// newPathTrie initializes the path trie.
+func newPathTrie(owner common.Hash, noLeftBound bool, db ethdb.KeyValueReader, batch ethdb.Batch) *pathTrie {
+	tr := &pathTrie{
+		owner:       owner,
+		noLeftBound: noLeftBound,
+		db:          db,
+		batch:       batch,
+	}
+	tr.tr = trie.NewStackTrie(tr.onTrieNode)
+	return tr
+}
+
+// onTrieNode is invoked whenever a new node is produced by the stackTrie.
```

Comment:
"node is produced" -- what does that mean? 
I take it to mean "a node (leaf or internal) is ready to be removed from the stacktrie and committed to the external sink".

Reply:
> exactly

Reply:
> exactly